fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,666 @@
+      return -ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      return -ErrorCodes.EFAULT();
+    }
+
+    return 0;
+  }
+
+  /**
+   * Opens an existing file for reading.
+   *
+   * Note that the open mode <emph>must</emph> be
+   * O_RDONLY, otherwise the open will fail. This is due to
+   * the Tachyon "write-once/read-many-times" file model.
+   *
+   * @param path the FS path of the file to open
+   * @param fi FileInfo data structure kept by FUSE
+   * @return 0 on success, a negative value on error
+   */
+  @Override
+  public int open(String path, FuseFileInfo fi) {
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    final int flags = fi.flags.get();
+    LOG.trace("open({}, 0x{}) [Tachyon: {}]", path, Integer.toHexString(flags), turi);
+
+    if ((flags & 3) != O_RDONLY.intValue()) {
+      LOG.error("Files can only be opened in O_RDONLY mode ({})", path);
+      return -ErrorCodes.EACCES();
+    }
+    try {
+      final TachyonFile tf = mTFS.openIfExists(turi);
+      if (tf == null) {
+        LOG.error("File {} does not exist", turi);
+        return -ErrorCodes.ENOENT();
+      }
+      final FileInfo tfi = mTFS.getInfo(tf);
+      if (tfi.isFolder) {
+        LOG.error("File {} is a directory", turi);
+        return -ErrorCodes.EISDIR();
+      }
+      final OpenFileEntry ofe =
+          new OpenFileEntry(mTFS.getInStream(tf), null);
+
+      synchronized (mOpenFiles) {
+        if (mOpenFiles.size() == MAX_OPEN_FILES) {
+          LOG.error("Cannot open {}: too many open files", turi);
+          return ErrorCodes.EMFILE();
+        }
+        mOpenFiles.put(mNextOpenFileId, ofe);
+        fi.fh.set(mNextOpenFileId);
+
+        // Assuming I will never wrap around (2^64 open files are quite a lot anyway)
+        mNextOpenFileId += 1;
+      }
+
+    } catch (FileDoesNotExistException e) {
+      LOG.debug("File does not exist {}", path, e);
+      return -ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("IOException on {}", path, e);
+      return -ErrorCodes.EIO();
+    } catch (TachyonException e) {
+      LOG.error("TachyonException on {}", path, e);
+      return -ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      return -ErrorCodes.EFAULT();
+    }
+
+    return 0;
+  }
+
+  /**
+   * Reads data from an open file.
+   * @param path the FS path of the file to read
+   * @param buf FUSE buffer to fill with data read
+   * @param size how many bytes to read
+   * @param offset offset of the read operation
+   * @param fi FileInfo data structure kept by FUSE
+   * @return the number of bytes read or 0 on EOF. A negative
+   *         value on error
+   */
+  @Override
+  public int read(String path, Pointer buf, @size_t long size, @off_t long offset,
+                  FuseFileInfo fi) {
+
+    if (size > Integer.MAX_VALUE) {
```

Comment:
It seems to me the size to read data is no more than 2GB (Integer.MAX_VALUE or  2147483647)? This may worth a documentation in the javadoc of this method.